crates/sui-types/src/effects/effects_v2.rs

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::base_types::{EpochId, ObjectDigest, ObjectRef, TransactionDigest};
+use crate::digests::TransactionEventsDigest;
+use crate::effects::{TransactionEffectsAPI, TransactionEffectsDebugSummary};
+use crate::execution_status::ExecutionStatus;
+use crate::gas::GasCostSummary;
+use crate::object::Owner;
+use crate::storage::{DeleteKind, WriteKind};
+use crate::{ObjectID, SequenceNumber};
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+
+/// The response from processing a transaction or a certified transaction
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+pub struct TransactionEffectsV2 {
+    /// The status of the execution
+    status: ExecutionStatus,
+    /// The epoch when this transaction was executed.
+    executed_epoch: EpochId,
+    gas_used: GasCostSummary,
+    /// The transaction digest
+    transaction_digest: TransactionDigest,
+    /// The updated gas object reference. Have a dedicated field for convenient access.
+    /// It's also included in `changed_objects`.
+    gas_object: (ObjectRef, Owner),
+    /// The digest of the events emitted during execution,
+    /// can be None if the transaction does not emit any event.
+    events_digest: Option<TransactionEventsDigest>,
+    /// The set of transaction digests this transaction depends on.
+    dependencies: Vec<TransactionDigest>,
+
+    /// The version number of all the written Move objects by this transaction.
+    lamport_version: SequenceNumber,
+    /// Objects whose state are changed in the object store.
+    changed_objects: Vec<(ObjectID, ObjectChange)>,
+    /// Shared objects that are not mutated in this transaction. Unlike owned objects,
+    /// read-only shared objects' version are not committed in the transaction,
+    /// and in order for a node to catch up and execute it without consensus sequencing,
+    /// the version needs to be committed in the effects.
+    unchanged_shared_objects: Vec<(ObjectID, UnchangedSharedKind)>,
```

Comment:
did you consider extending ObjectOut with a ReadOnlySharedObject enumerant? Then this could be merged into changed_objects

Reply:
Yes.
My concern with that direction is that it almost doesn't share any similarity to the changed_objects except each has a unique object ID. A few potential problems:
1. It creates a lot of impossible combinations, for example, id_created and id_deleted would never be true in this combo.
2. The logic of dealing with object changes become more complicated, for instance, for object pruning, I could simply prune everything whose ObjectIn is Exist; or for state accumulation, I could simply remove everything whose ObjectIn is Exist. But now with this new variant, I will have to filter out anything whose out is ReadOnly.